# users_items

In [1]:
from extract import jsongz

#df_original = jsongz.leer_archivo("..\\data\\raw\\users_items.json.gz")
df_original = jsongz.leer_archivo("../data/raw/users_items.json.gz")

In [2]:
df = df_original.copy()
df.head()

user_id  items_count           steam_id  \
0  76561197970982479          277  76561197970982479   
1            js41637          888  76561198035864385   
2          evcentric          137  76561198007712555   
3         Riot-Punch          328  76561197963445855   
4              doctr          541  76561198002099482   

                                            user_url  \
0  http://steamcommunity.com/profiles/76561197970...   
1               http://steamcommunity.com/id/js41637   
2             http://steamcommunity.com/id/evcentric   
3            http://steamcommunity.com/id/Riot-Punch   
4                 http://steamcommunity.com/id/doctr   

                                               items  
0  [{'item_id': '10', 'item_name': 'Counter-Strik...  
1  [{'item_id': '10', 'item_name': 'Counter-Strik...  
2  [{'item_id': '1200', 'item_name': 'Red Orchest...  
3  [{'item_id': '10', 'item_name': 'Counter-Strik...  
4  [{'item_id': '300', 'item_name': 'Day of Defea...

In [3]:
df.shape[0]

88310

En un principio, hay 88310 registros en el dataframe

Evaluar los tipos de datos usados por las columnas

In [4]:
from transform import info

info.mostrar_info_key_types(df)

user_id - {<class 'str'>}
items_count - {<class 'int'>}
steam_id - {<class 'str'>}
user_url - {<class 'str'>}
items - {<class 'list'>}


Recordar que las columnas que contiene valores del tipo "float", podrian indica que contine valores nulos.

Como la columna "items" es una lista.
Hay que desanidarlo para poder manejar los datos en su interior.

In [5]:
from transform import normalizar

df = normalizar.desanidar_dataframe(df, ["items"])

df.head()

user_id  items_count           steam_id  \
0  76561197970982479          277  76561197970982479   
1  76561197970982479          277  76561197970982479   
2  76561197970982479          277  76561197970982479   
3  76561197970982479          277  76561197970982479   
4  76561197970982479          277  76561197970982479   

                                            user_url  \
0  http://steamcommunity.com/profiles/76561197970...   
1  http://steamcommunity.com/profiles/76561197970...   
2  http://steamcommunity.com/profiles/76561197970...   
3  http://steamcommunity.com/profiles/76561197970...   
4  http://steamcommunity.com/profiles/76561197970...   

                                               items  
0  {'item_id': '10', 'item_name': 'Counter-Strike...  
1  {'item_id': '20', 'item_name': 'Team Fortress ...  
2  {'item_id': '30', 'item_name': 'Day of Defeat'...  
3  {'item_id': '40', 'item_name': 'Deathmatch Cla...  
4  {'item_id': '50', 'item_name': 'Half-Life: Opp...

Evaluamos nuevamente los tipos de datos

In [6]:
from transform import info

info.mostrar_info_key_types(df)

user_id - {<class 'str'>}
items_count - {<class 'int'>}
steam_id - {<class 'str'>}
user_url - {<class 'str'>}
items - {<class 'float'>, <class 'dict'>}


Ahora que se puede observar que el contenido de la columna "items" son datos de tipo "dict" y "float", hay que expandir la columna "items", ya que es de tipo "dict".

In [7]:
from transform import normalizar

df = normalizar.expandir_columna(df, "items")

df.head()

user_id  items_count           steam_id  \
0  76561197970982479          277  76561197970982479   
1  76561197970982479          277  76561197970982479   
2  76561197970982479          277  76561197970982479   
3  76561197970982479          277  76561197970982479   
4  76561197970982479          277  76561197970982479   

                                            user_url item_id  \
0  http://steamcommunity.com/profiles/76561197970...      10   
1  http://steamcommunity.com/profiles/76561197970...      20   
2  http://steamcommunity.com/profiles/76561197970...      30   
3  http://steamcommunity.com/profiles/76561197970...      40   
4  http://steamcommunity.com/profiles/76561197970...      50   

                   item_name  playtime_forever  playtime_2weeks  
0             Counter-Strike               6.0              0.0  
1      Team Fortress Classic               0.0              0.0  
2              Day of Defeat               7.0              0.0  
3         Deathmatch Classic               0.0              0.0  
4  Half-Life: Opposing Force               0.0              0.0

revisar nuevamente los tipos de datos usado en cada columna

In [8]:
from transform import info

info.mostrar_info_key_types(df)

user_id - {<class 'str'>}
items_count - {<class 'int'>}
steam_id - {<class 'str'>}
user_url - {<class 'str'>}
item_id - {<class 'str'>, <class 'float'>}
item_name - {<class 'str'>, <class 'float'>}
playtime_forever - {<class 'float'>}
playtime_2weeks - {<class 'float'>}


Las columnas que utilizan datos de tipo "float" , ademas de otros tipos, debe referirse a valores nulos

Se elimina las columnas que no son relevantes para los objetivos del proyecto. Esto está desarrollado en una hoja de cálculo, en una matriz de objetivos vs. datos.

https://docs.google.com/spreadsheets/d/1ccOxyOM_yctvpk4-CbpX4kCihNn1rX1QifKjbtov10s/edit?usp=sharing

In [9]:
from transform import limpiar

limpiar.eliminar_columnas_innecesarias(
    df, ["items_count", "steam_id", "user_url", "playtime_2weeks", "item_name"]
)

df.head()

user_id item_id  playtime_forever
0  76561197970982479      10               6.0
1  76561197970982479      20               0.0
2  76561197970982479      30               7.0
3  76561197970982479      40               0.0
4  76561197970982479      50               0.0

In [10]:
df.shape[0]

5170015

Ahora, la cantidad de filas en el dataframe es de: 5170015

## Variables
Las que utiliza este dataframe son:

### Variables categóricas
* user_id - variable categórica nominal (indentificador)
* item_id - variable categórica nominal (indentificador)
* playtime_forever - variable numérica

## Limpieza

Revisar la cantidad de celdas vacias en el dataframe

In [11]:
from transform import info

info.celdas_vacias(df)

Cantidad  Porcentaje
user_id                  0    0.000000
item_id              16806    0.003251
playtime_forever     16806    0.003251

Como "itemd_id" es un identificador, no puede ser nulo. Así que hay que eliminar las columnas donde es nulo.

In [12]:
from transform import limpiar

limpiar.eliminar_filas_por_celdas_vacias_en(df, ["item_id"])

la cantidad de filas en el dataframe es: 5170015
despues de eliminar las filas vacias es: 5153209


In [13]:
from transform import limpiar

limpiar.eliminar_filas_vacias(df)

la cantidad de filas en el dataframe es: 5153209
despues de eliminar las filas vacias es: 5153209


In [14]:
from transform import limpiar

df = limpiar.eliminar_filas_duplicadas(df)

la cantidad de filas en el dataframe es: 5153209
despues de eliminar las filas repetidas es: 5094092


Volver a revisar la cantidad de celdas vacias en el dataframe

In [15]:
from transform import info

info.celdas_vacias(df)

Cantidad  Porcentaje
user_id                  0         0.0
item_id                  0         0.0
playtime_forever         0         0.0

In [16]:
df.shape[0]

5094092

Ahora, la cantidad de filas en el dataframe es de: 5094092

## Load

In [17]:
df.to_parquet("../data/intermediate/users_items.parquet")